### ISSUE TITLE
"asynchronous generator is already running" on Python 3.8 when wrapping __anext__

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#9805]
|[<img alt="wRAR's avatar" src="https://avatars.githubusercontent.com/u/241039?s=50" width="50" height="50">](https://github.com/wRAR)| @wRAR reported|
|-|-|
|Trac ID|trac#9805|
|Type|defect|
|Created|2020-04-20 13:48:27Z|

I'm writing code that consumes async generators as deferreds, using ```Deferred.fromFuture(asyncio.ensure_future(gen.__anext__()))``` (I don't know if it's supported but it seems the only way to do this).

Some cases, or maybe all of them, only work on 3.7 but give an error on 3.8:

```
  File ".env/lib/python3.8/site-packages/twisted/internet/defer.py", line 824, in adapt
    extracted = result.result()
builtins.RuntimeError: anext(): asynchronous generator is already running
```

This may be caused by some 3.8 behavior changes related to those discussed in https://bugs.python.org/issue38559

I'm attaching a simple test case. ```asyncio.sleep``` may be replaces with e.g. some aiohttp calls.

I'm using Python 3.8.2 from Debian unstable.

Attachments:

* [asyncgen_testcase.py](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/f26/f26584ac447dfb264e44db699d42f44114e5978a/9e1135846f26b6ee8ac7818aceb51024d3d47a29.py) (714 bytes) - added by wRAR on 2020-04-20 13:48:50Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__9805 9805
type__defect defect
reporter__wRAR wRAR
priority__normal normal
milestone__Python_3_8_Support Python 3.8 Support
branch__ 
branch_author__ 
status__closed closed
resolution__invalid invalid
component__core core
keywords__None None
time__1587390507832077 1587390507832077
changetime__1594723138236371 1594723138236371
version__None None
owner__None None

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="wRAR's avatar" src="https://avatars.githubusercontent.com/u/241039?s=50" width="50" height="50">](https://github.com/wRAR)<a name="note_2"></a>|@wRAR commented|
|-|-|

I think I've found the root cause, the generator is awaited several times in parallel, which is apparently not supported anymore, similar to what's described in he Python issue linked above. The same error is reproducible with simple `asyncio.gather()` without Twisted. So looks like I should close this ticket and find a correct way to implement this.

**Comment by twisted-trac** [comment]
|[<img alt="wRAR's avatar" src="https://avatars.githubusercontent.com/u/241039?s=50" width="50" height="50">](https://github.com/wRAR)|@wRAR set status to `closed`|
|-|-|
